Viktor.horvath ch415 wyss airtabler

vhorvath opened the pull request:
- Incorporated and reviewed the different patches we developed to work around the bugs in `airtabler`.
- Added `air_get_all()` function to factor out the functionality from the DataDashboards
- Created tests, which should help checking if the new functions work

vhorvath committed:
House keeping preventing secret leaks

vhorvath committed:
Ensuring that offset is exposed

vhorvath committed:
Readability: guard pattern for `air_parse()`

vhorvath committed:
Added tests for air_get()

vhorvath committed:
Patched .bind_df() for embedded data.frames

vhorvath committed:
Created test for data.frame binding

vhorvath committed:
Corrected comments in tests

vhorvath committed:
Created air_get_all() function

vhorvath committed:
Added simple test for air_get_all()

vhorvath committed:
Updated package documentation

vhorvath committed:
Replaced `while` loop with `repeat`

vhorvath committed:
Brief Readme for the testing

vhorvath committed:
Added air_get(), air_get_all()

vhorvath committed:
Update Readme.md

Corrected small typos

vhorvath committed:
Removed unnecessary `next` from .bind_df()

Removed unnecessary `next` from .bind_df()
Fixed a typo

ranikay reviewed (approved):
Nice! I was able to get all tests to run 👍 Great documentation / instructions.

austinjpaul commented on R/airtabler.R:
```diff
@@ -97,9 +98,47 @@ air_get <- function(base, table_name, record_id = NULL,
+                          table_name,
+                          view = NULL,
+                          sortField = NULL,
+                          sortDirection = NULL) {
```
minor indentation issue?

vhorvath replied:
Yeah, I fixed it on my rig, but then I did not want to send another commit with this because I invited you for review already. I'll patch it here.

austinjpaul commented on R/airtabler.R:
```diff
@@ -256,13 +295,16 @@ air_validate <- function(res) {
+
+  # Add offset, if available
+  if(!is.null(res_obj$offset)) {
+    attr(res, "offset") <- res_obj$offset
```
Nice readability improvement here.

vhorvath replied:
Thanks! :)

austinjpaul commented on R/airtabler.R:
```diff
@@ -530,24 +572,29 @@ air_table_funs <- function(base, table_name) {
+    for (col in all_col_names) {
+      if (class(df[[col]]) == 'data.frame') {
+        row.names(df[[col]]) <- df$id # this takes care of the embedded data.frames
+        next
```
Could this just be turned into an `else if` instead of using `next`?

vhorvath replied:
This is a very interesting question Austin, made me think about why I chose this approach!

In terms of logic and computation cost, your suggestion makes total sense and the outcome would be equal!

However, I tend to side with the guard pattern as much as I can, mostly because of the benefits of using it in pipelines. Maybe I am abusing it a little...

I tend to process logic more easily this way because I can declare in my head that if a column is of a type `data.frame` that will be dealt with in that branch, and I skip to the next item in the loop.

Simultaneously, I also declare that whatever happens downstream in the loop won't affect the outcome, so I don't need to read it for the `data.frame` case.

I also tend to use `else`, `else if` and `switch` when there are comparisons of the same type because I don't need to think of a different value domain. However, the following if statement in the loop deals with the `value` rather than the `class` of the column. I know that this is a minor distinction...

This thinking comes from `Python` where there isn't a real `switch` structure, though you can do `elif`s.

Guard pattern also tends to flatten the code, and unlinks dependencies, which in this case is not that necessary, we're still using the same object.

austinjpaul commented on tests/01_air_get.R:
```diff
@@ -0,0 +1,31 @@
+
+    if ( i == 1 ) { 
+        if ( dim(x)[1] < 100 ) {
+            stop('Test air_get(): This test requires a table with at least 100 records, check .Renviron.')
```
I wonder if it would make sense to add a test to write a large table to the test base?
For future improvement, not necessarily now.

vhorvath replied:
I think the tests should be expanded significantly, and this would take `airtabler` to the next level. It should be done in a real testing fashion: upload a local table (possibly random content with multiple types), then download it in different ways and compare them.

It would also require testing of error handling, not sure if `airtabler` has any.

I need to learn how to do proper unit tests in `R`, though...

austinjpaul reviewed (approved):
Awesome work, Viktor, this looks superb! 

I wonder how much of this is worth pushing up to bergant/airtabler. I guess we'll see what happens to your open PR there and can perhaps follow up with additional features if they're open to it.

vhorvath committed:
Made stylistic changes

Updated style of argument lists for `air_get()` and `air_get_all()` to match that used in the rest of the code.

vhorvath commented:
Thanks @ranikay and @austinjpaul for the reviews and suggestions!
